File: ios/Packages/GemstoneServices/Sources/Perpetual/HyperliquidObserverService.swift
// Copyright (c). Gem Wallet. All rights reserved.

import GemstonePrimitives
import class Gemstone.GemPerpetualService
import protocol Gemstone.GemPerpetualServiceProtocol
import protocol Gemstone.GemPerpetualStreamServiceProtocol
import Foundation
import struct Gemstone.GemPerpetualConnection
import enum Gemstone.GemPerpetualSubscription
import Primitives
import WebSocketClient

public actor HyperliquidObserverService: PerpetualObservable {
    private let perpetualService: any GemPerpetualServiceProtocol
    private let webSocket: any WebSocketConnectable
    private let streamService: any GemPerpetualStreamServiceProtocol

    private var observeTask: Task<Void, Never>?
    private var currentWallet: Wallet?

    public let chartService: any ChartStreamable

    public init(
        webSocket: any WebSocketConnectable,
        perpetualService: any GemPerpetualServiceProtocol,
        streamService: any GemPerpetualStreamServiceProtocol,
        chartService: any ChartStreamable = ChartObserverService(),
    ) {
        self.webSocket = webSocket
        self.perpetualService = perpetualService
        self.streamService = streamService
        self.chartService = chartService
    }

    deinit {
        observeTask?.cancel()
    }

    // MARK: - Public API

    public func setup(for wallet: Wallet) async {
        await connect(for: wallet)
    }

    public func disconnect() async {
        guard observeTask != nil else { return }

        observeTask?.cancel()
        observeTask = nil
        currentWallet = nil

        await streamService.disconnected()
        await webSocket.disconnect()
    }

    public func subscribe(_ subscription: GemPerpetualSubscription) async throws {
        try await streamService.subscribe(subscription: subscription)
    }

    public func unsubscribe(_ subscription: GemPerpetualSubscription) async throws {
        try await streamService.unsubscribe(subscription: subscription)
    }

    @discardableResult
    public func update(for wallet: Wallet) async -> PerpetualAccountMode? {
        guard let address = wallet.hyperliquidAccount?.address else { return nil }
        do {
            return try await perpetualService.syncPositions(walletId: wallet.id, address: address)
        } catch {
            debugLog("HyperliquidObserver: update failed: \(error)")
            return nil
        }
    }

    // MARK: - Private

    private func connect(for wallet: Wallet) async {
        guard currentWallet?.id != wallet.id else { return }

        await disconnect()

        let connection: GemPerpetualConnection?
        do {
            connection = try await perpetualService.connection(wallet: wallet.json())
        } catch {
            debugLog("HyperliquidObserver: connection failed: \(error)")
            return
        }
        guard let connection, let mode = try? PerpetualAccountMode(connection.mode) else { return }

        currentWallet = wallet
        observeTask = Task { [weak self] in
            guard let self else { return }
            await observeConnection(walletId: wallet.id, address: connection.address, mode: mode)
        }
    }

    private func observeConnection(walletId: WalletId, address: String, mode: PerpetualAccountMode) async {
        for await event in await webSocket.connect() {
            guard !Task.isCancelled else { break }

            switch event {
            case .connected:
                await handleConnected(address: address, mode: mode)
            case let .message(data):
                await handle(data, walletId: walletId, mode: mode)
            case .disconnected:
                await streamService.disconnected()
            }
        }
    }

    private func handleConnected(address: String, mode: PerpetualAccountMode) async {
        do {
            try await streamService.connected(address: address, mode: mode.json())
        } catch {
            debugLog("HyperliquidObserver: subscribe failed: \(error)")
        }
    }

    private func handle(_ data: Data, walletId: WalletId, mode: PerpetualAccountMode) async {
        do {
            guard let candle = try await streamService.handle(walletId: walletId.id, mode: mode.json(), data: data) else { return }
            try await chartService.yield(Primitives.ChartCandleUpdate(candle))
        } catch {
            debugLog("HyperliquidObserver: handle message failed: \(error)")
        }
    }

}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/perpetual/HyperliquidObserverService.kt
package com.gemwallet.android.data.services.gemstone.perpetual

import android.util.Log
import com.gemwallet.android.application.perpetual.cases.PerpetualObserver
import com.gemwallet.android.data.services.gemstone.stream.WebSocketConnectable
import com.gemwallet.android.data.services.gemstone.stream.WebSocketEvent
import com.gemwallet.android.ext.runCatchingCancellable
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.ChartCandleUpdate
import com.wallet.core.primitives.PerpetualAccountMode
import com.wallet.core.primitives.Wallet
import com.wallet.core.primitives.WalletId
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChangedBy
import kotlinx.coroutines.launch
import uniffi.gemstone.GemPerpetualService
import uniffi.gemstone.GemPerpetualServiceInterface
import uniffi.gemstone.GemPerpetualStreamService
import uniffi.gemstone.GemPerpetualStreamServiceInterface
import uniffi.gemstone.GemPerpetualSubscription

class HyperliquidObserverService(
    private val observePerpetualWallet: ObservePerpetualWallet,
    private val perpetualService: GemPerpetualServiceInterface,
    private val streamService: GemPerpetualStreamServiceInterface,
    private val connection: WebSocketConnectable,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
) : PerpetualObserver {

    private val foreground = MutableStateFlow(false)

    private val chartFlow = MutableSharedFlow<ChartCandleUpdate>(
        extraBufferCapacity = CHART_BUFFER_CAPACITY,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )

    override val chartUpdates: Flow<ChartCandleUpdate> = chartFlow.asSharedFlow()

    init {
        scope.launch {
            combine(foreground, observePerpetualWallet()) { isForeground, wallet ->
                wallet?.takeIf { isForeground }
            }
                .distinctUntilChangedBy { it?.id?.id }
                .collectLatest { wallet ->
                    wallet ?: return@collectLatest
                    val connection = runCatchingCancellable { perpetualService.connection(wallet.toGem()) }
                        .onFailure { Log.e(TAG, "Perpetual connection failed", it) }
                        .getOrNull() ?: return@collectLatest
                    observeConnection(wallet.id, connection.address, connection.mode.toPrimitives())
                }
        }
    }

    fun start() {
        foreground.value = true
    }

    fun stop() {
        foreground.value = false
    }

    override fun subscribe(subscription: GemPerpetualSubscription) {
        scope.launch { send { streamService.subscribe(subscription) } }
    }

    override fun unsubscribe(subscription: GemPerpetualSubscription) {
        scope.launch { send { streamService.unsubscribe(subscription) } }
    }

    private suspend fun observeConnection(walletId: WalletId, address: String, mode: PerpetualAccountMode) {
        connection.connect().collect { event ->
            when (event) {
                WebSocketEvent.Connected -> send { streamService.connected(address, mode.toGem()) }
                is WebSocketEvent.Message -> onMessage(walletId, mode, event.text)
                WebSocketEvent.Disconnected -> streamService.disconnected()
            }
        }
    }

    private suspend fun onMessage(walletId: WalletId, mode: PerpetualAccountMode, text: String) {
        runCatchingCancellable { streamService.candleUpdate(walletId.id, mode.toGem(), text.encodeToByteArray()) }
            .onSuccess { candle -> candle?.toPrimitives()?.let { chartFlow.emit(it) } }
            .onFailure { Log.e(TAG, "Socket message error: ${text.take(MESSAGE_LOG_LIMIT)}", it) }
    }

    private suspend fun send(request: suspend () -> Unit) {
        runCatchingCancellable(request).onFailure { Log.e(TAG, "Subscription request error", it) }
    }

    companion object {
        private const val TAG = "HyperliquidObserver"
        private const val CHART_BUFFER_CAPACITY = 64
        private const val MESSAGE_LOG_LIMIT = 100
    }
}
